cloudfour/pwa-workshop
# Feature: Exercise 7, Push Notifications

## Overview

This PR adds Push notifications logic.

## Screenshots

![push-example](https://user-images.githubusercontent.com/459757/59087432-90651780-893f-11e9-864c-7ee9f95ef803.gif)

## Testing

1. Load up https://deploy-preview-28--festive-keller-9a29cb.netlify.com and confirm:
    - the "**Enable Push Messaging**" button is visible
    - there is a "Push: NOT subscribed" log in the console

1. In the DevTools, go to **Application** > **Service Workers** and click the **Push** button and confirm:
    - you'll see a log in the console: Service Worker: Push Received: <your text>
    - no notification should trigger

1. Click the **Enable Push Messaging** button, the Notifications permission dialog should open
    <img width="612" alt="Screen Shot 2019-06-07 at 4 22 44 PM" src="https://user-images.githubusercontent.com/459757/59087730-81329980-8940-11e9-8e17-52830a6e6800.png">

1. Click **Block** or close the dialog and confirm:
    - you see the following in the console:
      ```
      Push: Failed to subscribe the user: DOMException
      Push: Server delete subscription
      ```
    - the button now reads **Push Messaging Blocked** and is disabled

1. Refresh the page and confirm:
    - the button still reads **Push Messaging Blocked** and is disabled

1. Change your notification settings back to "**Ask**" to undo the block

    <img width="401" alt="Screen Shot 2019-06-07 at 4 26 39 PM" src="https://user-images.githubusercontent.com/459757/59087927-103fb180-8941-11e9-9205-7dd280e87ec9.png">

1. On page refresh, confirm:
    - the button is enabled 
    - the button reads **Enable Push Messaging**

1. Click the **Enable Push Messaging** button again to trigger the "Show notifications" dialog. This time, though, choose **Allow** and confirm:
    - the following is logged:
      ```
      Push: User is subscribed!!!
      Push: Server store new subscription: <subscription object>
      ```
    - the button now reads **Disable Push Message**

1. Refresh the page and confirm
    - the button still reads **Disable Push Message**
    - the log reads:
      ```
      Push: Subscribed
      ```

1. In the DevTools, go to **Application** > **Service Workers** and click the **Push** button again and confirm:
    - a notification is visible with the text in the "Push" field
    - clicking the Notification opens `cloudfour.com`

      <img width="468" alt="Screen Shot 2019-06-07 at 4 31 00 PM" src="https://user-images.githubusercontent.com/459757/59088204-c905f080-8941-11e9-9e22-360f27372d1f.png">

1. Click the **Disable Push Messaging** button and confirm
    - the console reads:
      ```
      Push: Server delete subscription
      Push: User is unsubscribed
      ```
    - the button now reads **Enable Push Messaging**
    
---

- [Trello Card](https://trello.com/c/bpwgV6Q0/49-service-worker-exercise-7-code)
